scim2-sdk-common/src/test/java/com/unboundid/scim2/common/AddOperationValueFilterTestCase.java

```diff
@@ -0,0 +1,288 @@
+    );
+    assertThatThrownBy(() -> applyPatchRequest(improperFormat, new UserResource()))
+        .isInstanceOf(BadRequestException.class)
+        .hasMessageContaining("cannot set the 'value' field to an array");
+
+    // Attempt using a value selection filter on an attribute that is
+    // single-valued.
+    UserResource userWithLanguage = new UserResource().setPreferredLanguage("nb-NO");
+    Path singleValuedAttr =
+        Path.fromString("preferredLanguage[type eq \"work\"].value");
+    PatchRequest invalidAttr = createAddRequest(singleValuedAttr, "en-US");
+    assertThatThrownBy(() -> applyPatchRequest(invalidAttr, userWithLanguage))
+        .isInstanceOf(BadRequestException.class)
+        .hasMessageContaining("could not be processed")
+        .hasMessageContaining("value selection filter was provided");
+
+    // Perform another add on a single-valued attribute when it does not have a
+    // pre-existing value.
+    PatchRequest invalidAttr2 = createAddRequest(singleValuedAttr, "en-US");
+    assertThatThrownBy(() -> applyPatchRequest(invalidAttr2, new UserResource()))
+        .isInstanceOf(JsonProcessingException.class);
+
+    // Value filters must be the first element in the path.
+    Path nestedFilter = Path.fromString("parent.examples[type eq \"best\"].value");
+    assertThatThrownBy(() -> createAddRequest(nestedFilter, "value"))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasCauseInstanceOf(BadRequestException.class)
+        .hasMessageContaining("Path cannot target sub-attributes");
+
+    // Providing an "empty" filter should not be allowed.
+    assertThatThrownBy(() -> {
+      Path emptyFilterPath = Path.fromString("entitlements[].value");
+      PatchRequest emptyFilterReq = createAddRequest(emptyFilterPath, "ent1");
+      applyPatchRequest(emptyFilterReq, new UserResource());
+    }).isInstanceOf(BadRequestException.class);
+  }
+
+  /**
+   * Test the appearance of a User resource after a patch request with a value
+   * selection filter has updated the resource. In particular, this test
+   * verifies that the attribute specified within the filter is listed second
+   * in the JSON. For example, for an attribute path of
+   * {@code emails[type eq "work"].value}, the {@code "type": "work"} field in
+   * the JSON should be printed after the {@code value} field:
+   * <pre>
+   *   {
+   *     "value": "sissel@example.com",
+   *     "type": "work"
+   *   }
+   * </pre>
+   *
+   * @throws Exception  If an unexpected error occurs.
+   */
+  @Test
+  public void testDeserializedObject() throws Exception
+  {
+    final String rawExpectedResult = "{"
+        + "  \"schemas\" : [ \"urn:ietf:params:scim:schemas:core:2.0:User\" ],"
+        + "  \"emails\" : [ {"
+        + "    \"value\" : \"sissel@example.com\","
+        + "    \"type\" : \"work\""
+        + "  }, {"
+        + "    \"value\" : \"sissel2@example.com\","
```

Comment:
I think "yomiel@example.com" could be used here.

Reply:
Excellent suggestion; done!